### PR TITLE
store failures in preferences

### PR DESCRIPTION
**Pull request by mnbogner**
The idea here was to minimize attention to methods that are working by not continually trying them alongside methods that have been previously been blocked.  Since the user has no direct control, I was worried about the potential of basically disabling the app unless the app data is cleared, so I thought a one week retry limit might be reasonable.

I do wonder how we should integrate this with DNSTT.  Would it make more sense to also remember whether DNSTT worked and decide whether or not to try it, or would DNSTT be a way to provide fresh methods to try when the app would otherwise be inactive for at least a week?

### CODE REVIEW

**Comment by vitriolix** [comment]
one week feels like a long time, how did you come to that value?

**Comment by mnbogner** [comment]
> one week feels like a long time, how did you come to that value?

my starting point was not retrying them at all, but that seemed like it could leave the user in a bad place.  my understanding is that if something is blocked it's not likely to be unblocked unless something changes the configuration of the network.

**Comment by eighthave** [comment]
I see the dnstt channel as a way to get updated configs for a) when the app has 
been inactive for a while, and b) when the existing methods stop working.  I 
think there needs to be some active refreshing to ensure that there are enough 
alternatives for the case when the existing methods are blocked.

**Comment by eighthave** [comment]
There are some standard algorithms for retrying things that could be useful 
here.  The basic idea is having retry cycles.  After each fail, that particular 
thing will be bumped to a longer retry interval.

**Comment by stevenmcdonald** [comment]
FYI, DNS over HTTPS and DNS over TCP (which are what DNSTT uses) both seem thoroughly blocked, so I think DNSTT is kind of a moot point unless or until that changes... or did we find something that works and I'm forgetting?

**Comment by mnbogner** [comment]
> FYI, DNS over HTTPS and DNS over TCP (which are what DNSTT uses) both seem thoroughly blocked, so I think DNSTT is kind of a moot point unless or until that changes... or did we find something that works and I'm forgetting?

is the method itself blocked or could it work if we tried a different doh/dot parameter?

**Comment by mnbogner** [comment]
> There are some standard algorithms for retrying things that could be useful here. The basic idea is having retry cycles. After each fail, that particular thing will be bumped to a longer retry interval.

do you have any thoughts about time frames?  i was working from the assumption that anything that was blocked would stay blocked (and that we might attract more attention by retrying blocked resources) so i don't think i'd want to start with the usual 10/20/30 second intervals.

**Comment by eighthave** [comment]
> do you have any thoughts about time frames?  i was working from the 
assumption that anything that was blocked would stay blocked (and that we might 
attract more attention by retrying blocked resources) so i don't think i'd want 
to start with the usual 10/20/30 second intervals.

I think it happens quite often that things are blocked for only a temporary time 
period.  Retrying blocked things after a while seems worthwhile, if it is easy.

**Comment by mnbogner** [comment]
> I think it happens quite often that things are blocked for only a temporary time period. Retrying blocked things after a while seems worthwhile, if it is easy.

i added some code to implement this behavior.

File: android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt [hunk 1]
```diff
@@ -186,21 +187,55 @@ class NetworkIntentService : IntentService("NetworkIntentService") {
             }
         }
 
-        if (urls.isNullOrEmpty()) {
+        val urlsToSubmit = mutableListOf<String>()
+        if (!urls.isNullOrEmpty()) {
+            urls.forEach { url ->
```
Comment (vitriolix) [inline android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt:192]: I think you can simplify this code with an elvis operator ?:

Comment (mnbogner) [reply android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt:192]: that only catches nulls, right?

Comment (vitriolix) [reply android/envoy/src/main/kotlin/org/greatfire/envoy/NetworkIntentService.kt:192]: good point

**Pull request merged**